src/Client.php

```diff
@@ -272,7 +273,126 @@ public function getAsset($assetId) {
+   */
+  protected function getPresignUrl(array $file_data) {
+    $this->checkAuth();
+
+    $response = $this->client->request(
+      "GET",
+      $this->baseUrl . '/ws/awss3/generateupload',
+      [
+        'headers' => $this->getDefaultHeaders(),
+        'query' => $file_data,
+      ]
+    );
+    return json_decode($response->getBody());
+  }
+
+  /**
+   * Uploads file to Webdam AWS S3.
+   *
+   * @param mixed $presignedUrl
+   *   The presigned URL we got in previous step from AWS.
+   * @param string $file_uri
+   *   The file URI.
+   * @param string $file_type
+   *   The File Content Type.
+   * @return array
+   *   Response Status 100 / 200
+   */
+  protected function uploadPresigned($presignedUrl, $file_uri, $file_type) {
+    $this->checkAuth();
+
+    $response = $this->client->request(
+      "PUT",
+      $presignedUrl, [
+        'headers' => ['Content-Type' => $file_type],
+        'body' => $file_uri,
+      ]);
+    return [
+      'status' => json_decode($response->getStatusCode(), TRUE),
+    ];
+
+  }
+
+  /**
+   * Confirms the upload to Webdam.
+   *
+   * @param string $pid
+   *   The Process ID we got in first step.
+   *
+   * @return string
+   *   The uploaded/edited asset ID.
+   */
+  protected function uploadConfirmed($pid) {
+    $this->checkAuth();
+
+    $response = $this->client->request(
+      "PUT",
+      $this->baseUrl . '/ws/awss3/finishupload/' . $pid,
+      ['headers' => $this->getDefaultHeaders()]
+    );
+
+    return (string) json_decode($response->getBody(), TRUE)['id'];
+
+  }
+
+  /**
+   * Uploads Assets to Webdam using the previously defined methods.
+   *
+   * @param array $file_data
+   *   The file data required by Webdam.
+   * @param int $folderID
+   *   The Webdam folder ID.
+   *
+   * @return array
+   *   Webdam response (processId, presign url, status, asset id).
+   */
+  public function uploadAsset(array $file_data, $folderID = NULL) {
+    $this->checkAuth();
+
+    if ($folderID != NULL) {
+      $file_data['folderid'] = $folderID;
+    }
+    $file_uri = $file_data['file_uri'];
+    $file_type = $file_data['contenttype'];
+    $response = [];
+    // Getting Pre-sign URL.
+    $presign = $this->getPresignUrl($file_data);
+    if (property_exists($presign, 'presignedUrl')) {
+      // Post-sign upload.
+      $postsign = $this->uploadPresigned($presign->presignedUrl, $file_uri, $file_type);
+      $response['processId'] = $presign->processId;
+      $response['presignUrl'] = $presign->presignedUrl;
+      $response['post_status'] = $postsign['status'];
+
+      if ($postsign['status'] == '200' || $postsign['status'] == '100') {
+        // Getting Asset ID.
+        $uploadConfirm = $this->uploadConfirmed($presign->processId);
+        $response['id'] = $uploadConfirm;
+      }
+      else {
+        $response['error'] = 'Failed to upload file after presigning.';
+      }
+    }
+    else {
+      $response['error'] = 'Failed to obtain presigned URL from Webdam.';
```

Comment:
Exception added. Although still not sure if its ok... I couldn't test it as I would expect in testUploadAssetFailed(). Maybe we can take a look into it later?